ledger-core/ledger/server/buildersvc/bundle/writer.go

```diff
@@ -0,0 +1,244 @@
+	}
+	return &bundleWriter{snap: snap}
+}
+
+var _ Writer = &bundleWriter{}
+
+// bundleWriter implements necessary synchronization guarantees for the snapshot writer.
+type bundleWriter struct {
+	mutex sync.Mutex
+	snap  SnapshotWriter
+
+	lastReady synckit.SignalChannel
+}
+
+func (p *bundleWriter) WaitWriteBundles(done synckit.SignalChannel, fn func(bool)) bool {
+	p.mutex.Lock()
+	prev, next := p._prepareWait(fn != nil)
+	p.mutex.Unlock()
+
+	if next == nil {
+		// there is nothing to wait: (1) no writers (2) all done (3) already rolled back
+		// and there is no fn to call
+		return true
+	}
+
+	ok := false
+	if prev != nil {
+		select {
+		case _, ok = <- prev:
+		case <- done:
+			// make sure that commit/rollback status will be propagated properly
+			go propagateReady(prev, next)
+			if fn != nil {
+				fn(false)
+			}
+			return false
+		}
+	}
+
+	switch {
+	case fn == nil:
+		if ok {
+			// propagate status properly as someone can be after this wait operation, hence can be affected by rollback
+			next <- struct{}{}
+		}
+		close(next)
+		return true
+
+	case ok:
+		defer func() {
+			next <- struct{}{}
+			close(next)
+		}()
+	default:
+		defer close(next)
+	}
+
+	fn(true)
+	return true
+}
+
+func (p *bundleWriter) MarkReadOnly() error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	return p.snap.MarkReadOnly()
+}
+
+func propagateReady(prev synckit.SignalChannel, next synckit.ClosableSignalChannel) {
+	_, ok := <- prev
+	if ok {
+		next <- struct{}{}
+	}
+	close(next)
+}
+
+func (p *bundleWriter) _prepareWait(alwaysSetNext bool) (prev synckit.SignalChannel, next synckit.ClosableSignalChannel) {
+	prev = p.lastReady
+	p.lastReady = nil
+
+	select {
+	case <- prev:
+		// ignore prev if can read it - it is either ok, or was rolled back completely
+		prev = nil
+	default:
+		// prev hasn't finished yet
+	}
+
+	if alwaysSetNext || prev != nil {
+		next = make(chan struct{}, 1)
+		p.lastReady = next
+	}
+
+	return prev, next
+}
+
+func (p *bundleWriter) WriteBundle(bundle Writeable, completedFn ResultFunc) error {
+	if completedFn == nil {
+		panic(throw.IllegalValue())
+	}
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	snapshot, err := p.snap.TakeSnapshot()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if snapshot != nil {
+			snapshot.Rollback(false)
+		}
+	}()
+
+	if err = bundle.PrepareWrite(snapshot); err != nil {
+		return err
+	}
+	if err = snapshot.Prepared(); err != nil {
+		return err
+	}
+
+	prev, next := p._prepareWait(true)
+	go p.applyBundleSafe(snapshot, bundle, prev, next, completedFn)
+
+	snapshot = nil
+	return nil
+}
+
+func (p *bundleWriter) applyBundleSafe(snapshot Snapshot, bundle Writeable,
+	prev synckit.SignalChannel, next synckit.ClosableSignalChannel, completedFn ResultFunc,
+) {
+	defer close(next) // to be executed as the very last one
+
+	locked := false
+	chained := false
+	rollback := true
+	var err error
+
+	defer func() {
+		switch {
+		case !rollback:
+			if locked {
+				p.mutex.Unlock()
```

Comment:
this is impossible case.

Reply:
it is, but this is a precaution for someone to try to touch this code